odoo/modules/packages.py

```diff
@@ -0,0 +1,148 @@
+        self.installed_version = None       # latest_version
+
+        # info for upgrade
+        self.upgrade_end = False            # if the package is need to upgrade end-xxx
+        self.old_version = None             # the version when added to packages
+
+        # dependencies
+        self.depends = set(manifest.get('depends', ()))
+        self.dependencies: List[Package] = []
+
+        self._depth_loop = False            # detects dependency loops
+
+    @tools.lazy_property
+    def depth(self) -> int:
+        """ Return the longest distance from self to module 'base' along dependencies. """
+        if self._depth_loop:
+            raise Exception('depends_loop')
+        self._depth_loop = True
+        if self.depends:
+            result = max(package.depth for package in self.dependencies) + 1
+        else:
+            assert self.name == 'base'
+            result = 0
+        self._depth_loop = False
+        return result
+
+    def sorting_key(self):
+        return (self.depth, self.name)
+
+    def demo_installable(self) -> bool:
+        return all(p.dbdemo for p in self.dependencies)
+
+
+class Packages:
+    """
+    Sorted Python packages for Odoo modules with the same names ordered by (package.depth, package.name)
+    """
+    def __init__(self, cr) -> None:
+        self._packages: Dict[str, Package] = {}
+        self._cr = cr
+
+    def __contains__(self, module: str) -> bool:
+        return module in self._packages
+
+    def __getitem__(self, module: str) -> Package:
+        return self._packages[module]
+
+    def __iter__(self) -> Iterator[Package]:
+        return iter(sorted(self._packages.values(), key=Package.sorting_key))
+
+    def __len__(self) -> int:
+        return len(self._packages)
+
+    def __bool__(self) -> bool:
+        return bool(self._packages)
+
+    def add(self, names: Iterable[str]) -> None:
+        for name in names:
+            package = self._packages[name] = Package(name)
+            if not package.manifest.get('installable'):
+                message = None if name == 'studio_customization' else 'not_installable'
+                self._remove(name, message)
+
+        self._update_dependencies(names)
+        self._check_dependencies(names)
+        self._update_from_database(names)
+
+    def _update_dependencies(self, names: Iterable[str]) -> None:
+        for name in names:
+            package = self._packages.get(name)
+            if package is not None:
+                try:
+                    package.dependencies = [self._packages[dep] for dep in package.depends]
+                except KeyError:
+                    self._remove(name, 'missing_depends')
+
+    def _check_dependencies(self, names: Iterable[str]) -> None:
+        for name in names:
+            package = self._packages.get(name)
+            try:
+                package and package.depth
+            except Exception:
+                self._remove(name, 'depends_loop')
```

Comment:
Maybe we can manually `raise RecursionError` in `depth()` instead of `raise Exception`
and
```py
except RecursionError:
    self._remove(name, 'depends_loop')
```
here

Reply:
That's fine by me, I wasn't sure about the exception type.
The point was to raise and exception, instead of removing the node and returning nonsensical values.